libp2p/Session.h

```diff
@@ -87,99 +88,103 @@ class SessionFace
 class Session: public SessionFace, public std::enable_shared_from_this<SessionFace>
 {
 public:
-	Session(Host* _server, std::unique_ptr<RLPXFrameCoder>&& _io, std::shared_ptr<RLPXSocket> const& _s, std::shared_ptr<Peer> const& _n, PeerSessionInfo _info);
-	virtual ~Session();
+    Session(Host* _server, std::unique_ptr<RLPXFrameCoder>&& _io, std::shared_ptr<RLPXSocket> const& _s, std::shared_ptr<Peer> const& _n, PeerSessionInfo _info);
+    virtual ~Session();
 
-	void start() override;
-	void disconnect(DisconnectReason _reason) override;
+    void start() override;
+    void disconnect(DisconnectReason _reason) override;
 
-	void ping() override;
+    void ping() override;
 
-	bool isConnected() const override { return m_socket->ref().is_open(); }
+    bool isConnected() const override { return m_socket->ref().is_open(); }
 
-	NodeID id() const override;
+    NodeID id() const override;
 
-	void sealAndSend(RLPStream& _s) override;
+    void sealAndSend(RLPStream& _s) override;
 
-	int rating() const override;
-	void addRating(int _r) override;
+    int rating() const override;
+    void addRating(int _r) override;
 
-	void addNote(std::string const& _k, std::string const& _v) override { Guard l(x_info); m_info.notes[_k] = _v; }
+    void addNote(std::string const& _k, std::string const& _v) override { Guard l(x_info); m_info.notes[_k] = _v; }
 
-	PeerSessionInfo info() const override { Guard l(x_info); return m_info; }
-	std::chrono::steady_clock::time_point connectionTime() override { return m_connect; }
+    PeerSessionInfo info() const override { Guard l(x_info); return m_info; }
+    std::chrono::steady_clock::time_point connectionTime() override { return m_connect; }
 
-	void registerCapability(CapDesc const& _desc, std::shared_ptr<Capability> _p) override;
+    void registerCapability(CapDesc const& _desc, std::shared_ptr<PeerCapabilityFace> _p) override;
 
-	std::map<CapDesc, std::shared_ptr<Capability>> const& capabilities() const override { return m_capabilities; }
+    std::map<CapDesc, std::shared_ptr<PeerCapabilityFace>> const& capabilities() const override
+    {
+        return m_capabilities;
```

Comment:
Why capabilities are shared? Cannot the Session be the only owner of them?

Reply:
Good question, I think it can, but then we don't have a guarantee that someone won't store reference to capability longer than `Session` lives.

Solution could be to remove capabilities getter and provide only the method to apply callback to the capabilities.